spec/test.spec.js

```diff
@@ -514,102 +625,144 @@ describe('S3Adapter tests', () => {
       options = {
         bucketPrefix: 'test/',
       };
+      s3ClientMock = jasmine.createSpyObj('S3Client', ['send']);
+      s3ClientMock.send.and.returnValue(Promise.resolve());
+    });
+
+    it('should save a file with right command', async () => {
+      const s3 = getMockS3Adapter(options);
+      s3._s3Client = s3ClientMock;
+
+      await s3.createFile('file.txt', 'hello world', 'text/utf8', {});
+      expect(s3ClientMock.send.calls.all().length).toBe(2);
+      expect(s3ClientMock.send).toHaveBeenCalledWith(jasmine.any(CreateBucketCommand));
+      expect(s3ClientMock.send).toHaveBeenCalledWith(jasmine.any(PutObjectCommand));
     });
 
     it('should save a file with metadata added', async () => {
       const s3 = getMockS3Adapter(options);
-      s3._s3Client.upload = (params, callback) => {
-        const { Metadata } = params;
-        expect(Metadata).toEqual({ foo: 'bar' });
-        const data = {
-          Body: Buffer.from('hello world', 'utf8'),
-        };
-        callback(null, data);
-      };
-      const fileName = 'randomFileName.txt';
+      s3._s3Client = s3ClientMock;
       const metadata = { foo: 'bar' };
-      await s3.createFile(fileName, 'hello world', 'text/utf8', { metadata });
+
+      await s3.createFile('file.txt', 'hello world', 'text/utf8', { metadata });
+      const commands = s3ClientMock.send.calls.all();
+      expect(commands.length).toBe(2);
+      expect(commands[0].args[0]).toBeInstanceOf(CreateBucketCommand);
+      const commandArg = commands[1].args[0];
+      expect(commandArg).toBeInstanceOf(PutObjectCommand);
+      expect(commandArg.input.Metadata).toEqual({ foo: 'bar' });
     });
 
     it('should save a file with tags added', async () => {
       const s3 = getMockS3Adapter(options);
-      s3._s3Client.upload = (params, callback) => {
-        const { Tagging } = params;
-        expect(Tagging).toEqual('foo=bar&baz=bin');
-        const data = {
-          Body: Buffer.from('hello world', 'utf8'),
-        };
-        callback(null, data);
-      };
-      const fileName = 'randomFileName.txt';
+      s3._s3Client = s3ClientMock;
       const tags = { foo: 'bar', baz: 'bin' };
-      await s3.createFile(fileName, 'hello world', 'text/utf8', { tags });
+
+      await s3.createFile('file.txt', 'hello world', 'text/utf8', { tags });
+      const commands = s3ClientMock.send.calls.all();
+      expect(commands.length).toBe(2);
+      expect(commands[0].args[0]).toBeInstanceOf(CreateBucketCommand);
+      const commandArg = commands[1].args[0];
+      expect(commandArg).toBeInstanceOf(PutObjectCommand);
+      expect(commandArg.input.Tagging).toBe('foo=bar&baz=bin');
     });
 
     it('should save a file with proper ACL with direct access', async () => {
-      // Create adapter
       options.directAccess = true;
       const s3 = getMockS3Adapter(options);
-      spyOn(s3._s3Client, 'upload').and.callThrough();
-      // Save file
+      s3._s3Client = s3ClientMock;
```

Comment:
Why are we using a mock client instead of spying? The issue with this is that we are now testing a mock client, which may not behave like the actual client if that changes, so the test becomes less robust.

Reply:
The issue here is not with the client itself, as I’ve consistently defined `const s3ClientMock = jasmine.createSpyObj('S3Client', ['send']);` and spied on the client to ensure realistic behavior.

What I believe you’re referring to is the handling of the adapter itself. I’ve been mindful of this and have aimed to minimize the use of mocks for the adapter, sticking to the original adapter whenever possible. Specifically, I’ve kept the mocked adapter only in tests that are focused exclusively on the logic within the adapter itself and not on the client.

Reply:
If I misunderstood your point and you’re referring to why we’re using a mock client instead of testing with the actual S3 client, it’s primarily because we don’t have the necessary environment variables set up to enable the client’s realistic behavior during testing.

If you believe having realistic behavior is essential, we could set up a fake S3 (like using LocalStack or a similar tool) and test against it. This would allow us to avoid defining mock behaviors and instead validate interactions with a closer-to-real implementation of the S3 client.

Reply:
I think it's fine as is, thanks for the details.